# **MÓDULO 15 - Exercício**
# Análise  - A segunda etapa da Pré Modelagem

# 1) O primeiro exercício é o de salvar a base que criaram na atividade do módulo anterior em csv e abrir ela neste arquivo.
Igual fizemos no início do módulo atual no início da primeira aula.

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import plotly.express as px

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving df_mod14_tratado.csv to df_mod14_tratado.csv


In [ ]:
df = pd.read_csv("df_mod14_tratado.csv", delimiter=',')

df.head(10)

customerID  genero  idoso  casado  dependentes  tempo_como_cliente  \
0  7590-VHVEG  Female      0    True        False                   1   
1  5575-GNVDE    Male      0   False        False                  34   
2  3668-QPYBK    Male      0   False        False                   2   
3  7795-CFOCW    Male      0   False        False                  45   
4  8191-XWSZG  Female      0   False        False                  52   
5  9959-WOFKT    Male      0   False         True                  71   
6  4190-MFLUW  Female      0    True         True                  10   
7  4183-MYFRB  Female      0   False        False                  21   
8  8779-QRDMV    Male      1   False        False                   1   
9  1680-VDCWW    Male      0    True        False                  12   

   servico_telefonico servico_internet servico_seguranca  suporte_tecnico  \
0               False              DSL                No            False   
1                True              DSL               Yes            False   
2                True              DSL               Yes            False   
3               False              DSL               Yes             True   
4                True               No                No             True   
5                True      Fiber optic               Yes            False   
6                True              DSL                No             True   
7                True      Fiber optic                No            False   
8               False              DSL                No            False   
9                True               No                No             True   

   streaming_tv   tipo_contrato           metodo_pagamento  pagamento_mensal  \
0         False  Month-to-month           Electronic check             29.85   
1         False        One year               Mailed check             56.95   
2         False  Month-to-month               Mailed check             53.85   
3         False        One year  Bank transfer (automatic)             71.45   
4          True        One year               Mailed check             71.45   
5          True        Two year  Bank transfer (automatic)             71.45   
6         False  Month-to-month    Credit card (automatic)             71.45   
7         False  Month-to-month           Electronic check             71.45   
8         False  Month-to-month           Electronic check             71.45   
9          True        One year  Bank transfer (automatic)             71.45   

   total_pago  churn  
0       29.85  False  
1     1889.50  False  
2      108.15   True  
3     1840.75  False  
4     1022.95  False  
5     7382.25  False  
6      528.35   True  
7     1862.90  False  
8       39.65   True  
9      202.25  False

# 2) Comece pela análise univariada:

A) Utilize a função describe no seu dataframe, veja os insights que consegue retirar.

B) Já é possível identificar variáveis com possíveis outliers? Se sim, quais?

C) Plot gráficos que considerar importante para completar sua análise univariada. (Lembrando que sua variável preditora é o churn). Não se esqueça de trazer insights de cada gráfico plotado. Utilize pelo menos 4 variáveis distintas.

D) Verifique se os dados das variáveis Booleanas são balanceados ou não.

In [ ]:
# A) Utilize a função describe no seu dataframe, veja os insights que consegue retirar.
df['churn'] = df['churn'].astype(int)
df['streaming_tv'] = df['streaming_tv'].astype(int)
df['suporte_tecnico'] = df['suporte_tecnico'].astype(int)

df.describe()

idoso  tempo_como_cliente  suporte_tecnico  streaming_tv  \
count  2488.000000         2488.000000      2488.000000   2488.000000   
mean      0.161576           32.352090         0.497588      0.597669   
std       0.368135           24.636885         0.500095      0.490467   
min       0.000000            0.000000         0.000000      0.000000   
25%       0.000000            8.000000         0.000000      0.000000   
50%       0.000000           29.000000         0.000000      1.000000   
75%       0.000000           56.000000         1.000000      1.000000   
max       1.000000           72.000000         1.000000      1.000000   

       pagamento_mensal   total_pago        churn  
count       2488.000000  2488.000000  2488.000000  
mean          66.342564  2290.087178     0.260048  
std           28.051718  2264.585643     0.438749  
min           18.400000    18.800000     0.000000  
25%           45.400000   402.337500     0.000000  
50%           71.450000  1402.900000     0.000000  
75%           87.712500  3867.162500     1.000000  
max          118.650000  8564.750000     1.000000

### Com base no describe realizado no nosso dataframe podemos chegar a algumas considerações:

1. **Variável *idoso* (Variável Binária)**
* Média: 0.161 - Indica que cerca de 16% dos clientes são idosos
* Std: 0.368 - Indica que a maior parte dos valores está perto de zero (não idosos)

2. **Variável *tempo_como_cliente* (Variável quantitativa)**
* Média: 32.35 - Os clientes têm, em média, 32 meses como tempo de relacionamento.
* Std: 24.64 - Indica grande variação no tempo de permanência dos clientes.
* Mínimo - 0 meses, possivelmente clientes recém-cadastrados.
* Quartis (25%, 50%, 75%):
  * 25% dos clientes estão há 8 meses ou menos.
  * 50% (mediana) estão há 29 meses.
  * 75% estão há até 56 meses.
* Máximo - O maior tempo de relacionamento é de 72 meses.

3. **Variável *suporte_tecnico* (Variável Binária)**
* Média: 0.49 - Indica que cerca de 50% dos clientes utilizam suporte técnico.

4. **Variável *pagamento_mensal* (Variável Binária)**
* Média: 0.59 - Indica que cerca de 60% dos clientes utilizam serviços de streaming.

5. **Variável *pagamento_mensal* (Variável quantitativa)**
* Média: 66.34 - O pagamento médio é 66.34.
* Std: 28.05 - Indica uma variação significativa entre os valores pagos.
* Mínimo - O menor pagamento mensal é 18.40.
* Quartis (25%, 50%, 75%):
  * 25% pagam até 45.40.
  * 50% (mediana) pagam 71.45.
  * 75% pagam até 87.71.
* Máximo - O maior pagamento mensal registrado é 118.65.

6. **Variável *total_pago* (Variável quantitativa)**
* Média: 2290.09 - O valor médio total pago pelos clientes é 2,290.09.
* Std: 2264.59 - Sugere alta dispersão, com alguns clientes pagando valores muito maiores que outros.
* Mínimo - O menor total pago é 18.80, indicando clientes recentes ou inativos.
* Quartis (25%, 50%, 75%):
  * 25% pagaram até 402.34.
  * 50% (mediana) pagaram 1402.90.
  * 75% pagaram até 3867.16.
* Máximo - O maior total pago é 8,564.75, mostrando que há clientes de longa data com pagamentos altos.


Essas informações nos trazem os seguintes indícios:
* O tempo como cliente é bastante diverso, mas a maioria se encontra abaixo de 56 meses, indicando uma dificuldade de retenção por mais de 5 anos.
* Os pagamentos mensais têm uma distribuição assimétrica, com a maioria pagando valores abaixo de 87.71. É possível que existam planos mais populares nessa faixa de preço.
* O valor total pago é altamente variável.

B) Já é possível identificar variáveis com possíveis outliers? Se sim, quais?

Sim, o maior valor *total_pago* de 8564.75 indica a possibildiade de existência de outliers quando consideramos o valor médio pago de 2290.09.

In [ ]:
# B) Já é possível identificar variáveis com possíveis outliers? Se sim, quais?

# Variável tempo_como_cliente ----------------------------------------------------------------------------

# Calculando os quartis e IQR para a variável 'total_pago'
Q1 = df['total_pago'].quantile(0.25)
Q3 = df['total_pago'].quantile(0.75)
IQR = Q3 - Q1

# Determinando os limites inferior e superior
limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

# Identificando os outliers
outliers = df[(df['total_pago'] < limite_inferior) | (df['total_pago'] > limite_superior)]

# Exibindo os outliers encontrados
print('Variável tempo_como_cliente ---------------------------')
outliers

Variável tempo_como_cliente ---------------------------


Empty DataFrame
Columns: [customerID, genero, idoso, casado, dependentes, tempo_como_cliente, servico_telefonico, servico_internet, servico_seguranca, suporte_tecnico, streaming_tv, tipo_contrato, metodo_pagamento, pagamento_mensal, total_pago, churn]
Index: []

In [ ]:
# Variável pagamento_mensal ----------------------------------------------------------------------------

# Calculando os quartis e IQR para a variável 'pagamento_mensal'
Q1 = df['pagamento_mensal'].quantile(0.25)
Q3 = df['pagamento_mensal'].quantile(0.75)
IQR = Q3 - Q1

# Determinando os limites inferior e superior
limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

# Identificando os outliers
outliers = df[(df['pagamento_mensal'] < limite_inferior) | (df['pagamento_mensal'] > limite_superior)]

# Exibindo os outliers encontrados
print('Variável pagamento_mensal ---------------------------')
outliers

Variável pagamento_mensal ---------------------------


Empty DataFrame
Columns: [customerID, genero, idoso, casado, dependentes, tempo_como_cliente, servico_telefonico, servico_internet, servico_seguranca, suporte_tecnico, streaming_tv, tipo_contrato, metodo_pagamento, pagamento_mensal, total_pago, churn]
Index: []

In [ ]:
# Variável total_pago ----------------------------------------------------------------------------

# Calculando os quartis e IQR para a variável 'total_pago'
Q1 = df['total_pago'].quantile(0.25)
Q3 = df['total_pago'].quantile(0.75)
IQR = Q3 - Q1

# Determinando os limites inferior e superior
limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

# Identificando os outliers
outliers = df[(df['total_pago'] < limite_inferior) | (df['total_pago'] > limite_superior)]

# Exibindo os outliers encontrados
print('Variável total_pago ---------------------------')
outliers

Variável total_pago ---------------------------


Empty DataFrame
Columns: [customerID, genero, idoso, casado, dependentes, tempo_como_cliente, servico_telefonico, servico_internet, servico_seguranca, suporte_tecnico, streaming_tv, tipo_contrato, metodo_pagamento, pagamento_mensal, total_pago, churn]
Index: []

Analisando as variáveis que teriam a possibilidade de conter outliers, não identifiquei nenhum outlier em nenhuma das três. Levando em consideração o df tratado, conforme executado no módulo da pré modelagem.

In [ ]:
# C) Plot gráficos que considerar importante para completar sua análise univariada. (Lembrando que sua variável preditora é o churn). Não se esqueça de trazer insights de cada gráfico plotado. Utilize pelo menos 4 variáveis distintas.

# Boxplot para 'tempo_como_cliente'
fig_tempo = px.box(
    df,
    y='tempo_como_cliente',
    title='Box Plot do Tempo Como Cliente',
    points="all",  # Mostra os outliers
    labels={'tempo_como_cliente': 'Valores'}
)
fig_tempo.update_layout(
    title=dict(font=dict(size=16, family="Arial", color="darkblue")),
    yaxis=dict(title_font=dict(size=14, color="darkblue")),
    template="plotly_white",
)
fig_tempo.show()

# Boxplot para 'pagamento_mensal'
fig_pagamento = px.box(
    df,
    y='pagamento_mensal',
    title='Box Plot do Pagamento Mensal',
    points="all",
    labels={'pagamento_mensal': 'Valores'}
)
fig_pagamento.update_layout(
    title=dict(font=dict(size=16, family="Arial", color="darkblue")),
    yaxis=dict(title_font=dict(size=14, color="darkblue")),
    template="plotly_white",
)
fig_pagamento.show()

# Boxplot para 'total_pago'
fig_total = px.box(
    df,
    y='total_pago',
    title='Box Plot do Total Pago',
    points="all",
    labels={'total_pago': 'Valores'}
)
fig_total.update_layout(
    title=dict(font=dict(size=16, family="Arial", color="darkblue")),
    yaxis=dict(title_font=dict(size=14, color="darkblue")),
    template="plotly_white",
)
fig_total.show()



# 3) Identifique e trate as colunas que contém outliers.
Caso opte por mante-los ou altera-los justifique sua escolha.


In [ ]:
# Nesse caso não há o que tratar, não identifiquei nenhum outlier nas minhas variáveis. Acredito já ter feito o tratamento dos outliers no módulo anterior.

# 4) Realize a etapa da análise bivariada:
A) Questione pelo menos 5 informações e traga as respostas utilizando visuais gráficos e insights.


B) Quais variáveis você acredita serem as mais importantes para esse projetos relacionadas a variável Churn?

In [ ]:
# 1. Clientes mais antigos têm maior probabilidade de churn?
fig_tempo_churn = px.box(
    df,
    x='churn',
    y='tempo_como_cliente',
    color='churn',
    title='Tempo Como Cliente vs Churn',
    labels={'tempo_como_cliente': 'Tempo Como Cliente (meses)', 'churn': 'Churn'}
)
fig_tempo_churn.show()

### 1. Clientes mais antigos têm maior probabilidade de churn?

* Clientes sem churn (0):

  * Apresentam maior tempo médio como clientes, com a mediana em torno de 40 meses.
  * A distribuição é ampla, com alguns clientes alcançando até 70 meses de relação com a empresa.

* Clientes com churn (1):

  * Têm um tempo médio como clientes significativamente menor, com a mediana em torno de 20 meses.
  * A maioria dos clientes com churn tem um tempo como cliente menor que os sem churn.
* **Conclusão**:
Clientes mais antigos (com maior tempo de contrato) parecem ter menor probabilidade de churn. Isso pode indicar maior fidelidade ou satisfação conforme o relacionamento com a empresa se alonga. É importante investigar o que motiva clientes com menos tempo a cancelar, pois eles podem estar enfrentando problemas que podem ser resolvidos.

In [ ]:
# 2. Clientes com pagamento mensal mais alto apresentam maior churn?
fig_pagamento_churn = px.box(
    df,
    x='churn',
    y='pagamento_mensal',
    color='churn',
    title='Pagamento Mensal vs Churn',
    labels={'pagamento_mensal': 'Pagamento Mensal (R$)', 'churn': 'Churn'}
)
fig_pagamento_churn.show()

### 2. Clientes com pagamento mensal mais alto apresentam maior churn?

* Clientes com Churn (1):
  * Para os clientes que saem, a mediana do pagamento mensal é maior do que a mediana para os que permanecem. O IQR também está mais concentrado em valores mais altos para esses clientes.
* Clientes sem Chrun (0):
  * Para os clientes que permanecem, o IQR está em níveis mais baixos de pagamento mensal, e a mediana reflete uma tendência para valores menores em comparação com os clientes que saem.
* **Conclusão**: Clientes com pagamentos mensais mais altos têm maior probabilidade de churn. Isso pode indicar que esses clientes podem sentir que o custo é muito elevado para os benefícios percebidos. Estratégias para retenção, como descontos ou pacotes de valor agregado, poderiam ajudar a reduzir o churn nesse grupo.

In [ ]:
# 3. Clientes com suporte técnico ativo são menos propensos a churn?
fig_suporte_churn = px.histogram(
    df,
    x='suporte_tecnico',
    color='churn',
    barmode='group',
    title='Suporte Técnico Ativo vs Churn',
    labels={'suporte_tecnico': 'Suporte Técnico Ativo (0=Não, 1=Sim)', 'churn': 'Churn'},
    text_auto=True
)
fig_suporte_churn.show()

### 3. Clientes com suporte técnico ativo são menos propensos a churn?

* Clientes sem suporte técnico ativo (Suporte Técnico Ativo = 0):

  * Há uma quantidade significativa de churn entre os clientes que não têm suporte técnico ativo. Dos 1250 clientes nessa categoria, 525 (42%) realizaram churn.
  * Este grupo apresenta um maior risco de churn em comparação aos clientes com suporte técnico ativo.

* Clientes com suporte técnico ativo (Suporte Técnico Ativo = 1):

  * Apenas 122 clientes (10%) entre os 1238 com suporte técnico ativo realizaram churn.
  * A grande maioria dos clientes com suporte técnico ativo permaneceu.

* Impacto do suporte técnico ativo:

  * Clientes com suporte técnico ativo têm uma probabilidade significativamente menor de churn, o que sugere que o suporte técnico pode ser um fator importante para retenção.

* **Conclusão**: A presença de suporte técnico ativo parece reduzir consideravelmente o churn. Isso pode indicar que os clientes que recebem suporte percebem maior valor no serviço e estão mais satisfeitos. Para reduzir ainda mais o churn, seria interessante explorar a extensão do suporte técnico a clientes atualmente sem acesso.









In [ ]:
# 4. Clientes que utilizam streaming de TV têm maior churn?
fig_streaming_churn = px.histogram(
    df,
    x='streaming_tv',
    color='churn',
    barmode='group',
    title='Streaming TV vs Churn',
    labels={'streaming_tv': 'Streaming TV (0=Não, 1=Sim)', 'churn': 'Churn'},
    text_auto=True
)
fig_streaming_churn.show()

### 4. Clientes que utilizam streaming de TV têm maior churn?

* **Conclusões**:
  * Menor taxa de churn para quem possui Streaming TV: Clientes que têm Streaming TV apresentam uma taxa de cancelamento menor (21%) em comparação com os que não possuem (33%).
  * Base maior de clientes com Streaming TV: A quantidade de clientes que possuem Streaming TV é significativamente maior do que aqueles que não possuem. Isso pode indicar que o serviço de Streaming TV é popular entre os usuários.
  * Potencial impacto positivo do serviço de Streaming TV: A oferta de Streaming TV pode estar associada à retenção de clientes, sugerindo que ele pode ser um fator de fidelização.


In [19]:
# 5. Clientes que pagaram mais ao longo do tempo têm maior churn?
fig_total_pago_churn_boxplot = px.box(
    df,
    x='churn',
    y='total_pago',
    color='churn',
    title='Total Pago vs Churn',
    labels={'total_pago': 'Total Pago (R$)', 'churn': 'Churn'},
    points='all'  # Inclui todos os pontos no boxplot
)
fig_total_pago_churn_boxplot.show()


### 5. Clientes que pagaram mais ao longo do tempo têm maior churn?

* Clientes que não cancelaram (Churn = 0):

  * Esses clientes tendem a ter uma distribuição mais ampla de valores pagos, com alguns pagando valores significativamente mais altos.
  * A mediana do total pago é mais elevada em comparação com os clientes que cancelaram, sugerindo que clientes mais antigos (ou que gastaram mais) tendem a permanecer.
  * Existe uma maior dispersão nos valores, indicando uma variedade de perfis de consumo.

* Clientes que cancelaram (Churn = 1):

  * Os valores pagos são, em sua maioria, mais baixos, com menor dispersão.
  * A concentração de valores baixos sugere que muitos clientes cancelam após pouco tempo de uso ou com baixo envolvimento financeiro.
  * A presença de poucos valores mais altos indica que até mesmo clientes que gastaram mais podem cancelar, mas isso ocorre com menor frequência.

* Conclusões:
  * Clientes que gastaram mais tendem a ser mais leais, enquanto aqueles com pagamentos menores são mais propensos ao churn.
  * Estratégias de retenção podem focar em aumentar o valor percebido para clientes de menor gasto, incentivando-os a continuar.
  * Ofertas personalizadas podem ser criadas com base no perfil de pagamento para reduzir cancelamentos.